dataplane/handlers/interface.go

```diff
@@ -213,61 +213,102 @@ func (ni *Interface) reconcile(config *oc.Interface) {
 				log.Warningf("Failed to set mac address of port: %v", err)
-				return
+			}
+		}
+	} else {
+		if state.GetEthernet().GetHwMacAddress() != state.GetEthernet().GetMacAddress() {
+			log.V(1).Infof("resetting interface %s hw-addr %q", tapName, state.GetEthernet().GetHwMacAddress())
+			if err := kernel.SetInterfaceHWAddr(config.GetName(), state.GetEthernet().GetHwMacAddress()); err != nil {
+				log.Warningf("Failed to set mac address of port: %v", err)
 			}
 		}
 	}
+
 	if config.GetOrCreateSubinterface(0).Enabled != nil {
 		if state.GetOrCreateSubinterface(0).Enabled == nil || config.GetSubinterface(0).GetEnabled() != state.GetSubinterface(0).GetEnabled() {
 			log.V(1).Infof("setting interface %s enabled %t", tapName, config.GetSubinterface(0).GetEnabled())
 			if err := kernel.SetInterfaceState(tapName, config.GetSubinterface(0).GetEnabled()); err != nil {
 				log.Warningf("Failed to set state address of port: %v", err)
-				return
 			}
 		}
 	}
-	// TODO: refactor this.
-	for _, addr := range config.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
-		configIP := addr.Ip
-		configPL := addr.PrefixLength
-		var stateIP *string
-		var statePL *uint8
-		stateAddr := state.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp())
-		if stateAddr != nil {
-			stateIP = stateAddr.Ip
-			statePL = stateAddr.PrefixLength
-		}
-
-		if configIP != nil && configPL != nil && (stateIP == nil || *statePL != *configPL) {
-			log.V(1).Infof("Config IP: %v, Config PL: %v. State IP: %v, State PL: %v", addr.GetIp(), addr.GetPrefixLength(), stateAddr.GetIp(), stateAddr.GetPrefixLength())
-			log.V(2).Infof("setting interface %s ip %s/%d", tapName, *configIP, *configPL)
-			if err := kernel.SetInterfaceIP(tapName, *configIP, int(*configPL)); err != nil {
+
+	type addrPair struct {
+		cfgIP, stateIP *string
+		cfgPL, statePL *uint8
+	}
+
+	// Get all state IPs and their corresponding config IPs (if they exist).
+	var maybeDeletedPairs []*addrPair
+	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv4().Address {
+		pair := &addrPair{
+			stateIP: addr.Ip,
+			statePL: addr.PrefixLength,
+		}
+		if pairAddr := config.GetSubinterface(0).GetIpv4().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.cfgIP = pairAddr.Ip
+			pair.cfgPL = pairAddr.PrefixLength
+		}
+		maybeDeletedPairs = append(maybeDeletedPairs, pair)
+	}
+	for _, addr := range state.GetOrCreateSubinterface(0).GetOrCreateIpv6().Address {
+		pair := &addrPair{
+			stateIP: addr.Ip,
+			statePL: addr.PrefixLength,
+		}
+		if pairAddr := config.GetSubinterface(0).GetIpv6().GetAddress(addr.GetIp()); pairAddr != nil {
+			pair.cfgIP = pairAddr.Ip
+			pair.cfgPL = pairAddr.PrefixLength
+		}
+		maybeDeletedPairs = append(maybeDeletedPairs, pair)
+	}
+
+	for _, pair := range maybeDeletedPairs {
+		// If an IP exists in state, but not in config, remove the IP.
+		if pair.stateIP != nil && pair.statePL != nil && (pair.cfgIP == nil && pair.cfgPL == nil) {
```

Comment:
brackets around the state variables. Also I'm wondering if it should be `||` on the state and the config.

Reply:
i wanted to make sure that the config is completely gone or completely added before making any change